plugins/cros_ec-usb/fu-cros_ec-usb-device.c

```diff
@@ -0,0 +1,327 @@
+				continue;
+			ep = g_ptr_array_index (endpoints, 0);
+			self->iface_idx = g_usb_interface_get_number (intf);
+			self->ep_num = g_usb_endpoint_get_address (ep) & 0x7f;
+			self->chunk_len = g_usb_endpoint_get_maximum_packet_size (ep);
+
+			return TRUE;
+		}
+	}
+	g_set_error_literal (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_NOT_FOUND,
+			     "No Google Update interface found.");
+	return FALSE;
+}
+
+static gboolean
+fu_cros_ec_usb_device_open (FuUsbDevice *device, GError **error)
+{
+	GUsbDevice *usb_device = fu_usb_device_get_dev (device);
+	FuCrOSECUsbDevice *self = FU_CROS_EC_USB_DEVICE (device);
+
+	if (!g_usb_device_claim_interface (usb_device, self->iface_idx,
+					   G_USB_DEVICE_CLAIM_INTERFACE_BIND_KERNEL_DRIVER,
+					   error)) {
+		g_prefix_error (error, "failed to claim interface: ");
+		return FALSE;
+	}
+
+	/* success */
+	return TRUE;
+}
+
+
+static gboolean
+fu_cros_ec_usb_device_probe (FuUsbDevice *device, GError **error)
+{
+	FuCrOSECUsbDevice *self = FU_CROS_EC_USB_DEVICE (device);
+
+	/* very much like usb_updater2's usb_findit() */
+
+	if (!fu_cros_ec_usb_device_find_interface (device, error)) {
+		g_prefix_error (error, "failed to find update interface: ");
+		return FALSE;
+	}
+
+	if (0 == self->chunk_len) {
+		g_set_error (error,
+			     G_IO_ERROR,
+			     G_IO_ERROR_INVALID_DATA,
+			     "wMaxPacketSize isn't valid: %" G_GUINT16_FORMAT,
+			     self->chunk_len);
+		return FALSE;
+	}
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+do_xfer (FuCrOSECUsbDevice * cros_ec_device, guint8 *outbuf, gsize outlen,
+	 guint8 *inbuf, gsize inlen, gboolean allow_less,
+	 gsize *rxed_count, GError **error)
+{
+	GUsbDevice *usb_device = fu_usb_device_get_dev (FU_USB_DEVICE (cros_ec_device));
+	FuCrOSECUsbDevice *self = cros_ec_device;
+	gsize actual = 0;
+
+	/* send data out */
+	if (outbuf != NULL && outlen > 0) {
+		if (!g_usb_device_bulk_transfer (usb_device, self->ep_num,
+						 outbuf, outlen,
+						 &actual, 2000, NULL, error)) {
+			return FALSE;
+		}
+		if (actual != outlen) {
+			g_set_error (error,
+				     G_IO_ERROR,
+				     G_IO_ERROR_PARTIAL_INPUT,
+				     "only sent %" G_GSIZE_FORMAT "/%"
+				     G_GSIZE_FORMAT " bytes\n",
+				     actual, outlen);
+			return FALSE;
+		}
+	}
+
+	/* read reply back */
+	if (inbuf != NULL && inlen > 0) {
+		actual = 0;
+		if (!g_usb_device_bulk_transfer (usb_device,
+						 self->ep_num | 0x80,
+						 inbuf, inlen,
+						 &actual, 5000, NULL, error)) {
+			return FALSE;
+		}
+		if ((actual != inlen) && !allow_less) {
+			g_set_error (error,
+				     G_IO_ERROR,
+				     G_IO_ERROR_PARTIAL_INPUT,
+				     "only received %" G_GSIZE_FORMAT "/%"
+				     G_GSIZE_FORMAT " bytes\n",
+				     actual, outlen);
+			return FALSE;
+		}
+	}
+
+	if (rxed_count)
+		*rxed_count = actual;
+
+	return TRUE;
+}
+
+static gboolean
+fu_cros_ec_usb_device_setup (FuDevice *device, GError **error)
+{
+	FuCrOSECUsbDevice *self = FU_CROS_EC_USB_DEVICE (device);
```

Comment:
Maybe also `FuCrOSECUsbDevice` -> `FuCrosEcUsbDevice` to match convention.

Reply:
Ok, will do. Sorry for the weird CamelCase.

Reply:
No worries :) Maybe FuCrosEcDevice if we drop the USB everywhere too.

Reply:
Done.